mono/metadata/reflection.c

```diff
@@ -8390,19 +8390,17 @@ MonoCustomAttrInfo*
 mono_custom_attrs_from_class (MonoClass *klass)
 {
 	MonoDomain* domain = mono_domain_get();
-
-	// Hashing attributes as a lookup optimization.
-	MonoCustomAttrInfo* hashed_attribute_info = (MonoCustomAttrInfo*)g_hash_table_lookup(domain->class_custom_attributes, klass);
+	MonoCustomAttrInfo* cattrs;
+	MonoCustomAttrInfo* cattrs_orig;
 	guint32 idx;
 
-	if(hashed_attribute_info != NULL) 
-	{
-		return hashed_attribute_info;
-	}
-
 	if (klass->generic_class)
 		klass = klass->generic_class->container_class;
 
+	cattrs_orig = InterlockedCompareExchangePointer(&klass->cattrs, NULL, NULL);
```

Comment:
Could you get away with a non-atomic load here, or possibly in addition to the atomic, as an early out?  There's a fence when it's written so it seems like it would be safe.

Reply:
I'll defer to @joshpeterson who is a bit more knowledgeable on that topic.

Reply:
That said this is old mono in maintenance mode. I'll prefer correctness over performance.

Reply:
@corngood: We probably could get away with it, but I agree we should prefer correctness here, so I'm partial to keep this as-is.

Reply:
@joshpeterson probably a good idea